iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageInstallationCacheCheck.java

```diff
@@ -0,0 +1,191 @@
+
+@Rule(key = "S6587")
+public class PackageInstallationCacheCheck implements IacCheck {
+
+  private static final String MESSAGE = "Remove cache after installing packages.";
+
+  private static final Set<String> APK_CACHE_LOCATIONS = Set.of("/etc/apk/cache/*", "/var/cache/apk/*");
+  private static final Set<String> APT_COMMANDS = Set.of("apt", "apt-get", "aptitude");
+  private static final Set<String> APT_CACHE_LOCATION = Set.of("/var/lib/apt/lists/*");
+  private static final Predicate<String> containsROrF = s -> s.toLowerCase(Locale.ROOT).contains("r") || s.contains("f");
+  private static final Predicate<String> isFlag = s -> s.startsWith("-");
+  private static final Predicate<String> flagWithNecessaryRmOptions = isFlag.and(containsROrF);
+  private static final Predicate<String> flagWithoutContainingNecessaryFlags = isFlag.and(not(flagWithNecessaryRmOptions));
+
+  private static final CommandDetector APK_ADD = CommandDetector.builder()
+    .with("apk")
+    .withAnyFlagExcept("--no-cache")
+    .with("add")
+    .withAnyFlagExcept("--no-cache")
+    .build();
+
+  private static final CommandDetector APK_CLEAN = buildCleanCacheDetector(Set.of("apk"));
+  private static final CommandDetector REMOVE_APK_CACHE_DETECTOR = buildRemoveCacheDetector(APK_CACHE_LOCATIONS);
+  private static final CommandDetector APT_INSTALL = CommandDetector.builder()
+    .with(APT_COMMANDS)
+    .withAnyFlag()
+    .with("install")
+    .withAnyFlag()
+    .build();
+
+  private static final CommandDetector APT_CLEAN = buildCleanCacheDetector(APT_COMMANDS);
+  private static final CommandDetector REMOVE_APT_CACHE_DETECTOR = buildRemoveCacheDetector(APT_CACHE_LOCATION);
+
+  private static CommandDetector buildCleanCacheDetector(Set<String> commandNames) {
+    var builder = CommandDetector.builder()
+      .with(commandNames);
+
+    if (commandNames.contains("apk")) {
+      builder.with("cache");
+    }
+
+    return builder
+      .with("clean")
+      .build();
+  }
+
+  private static CommandDetector buildRemoveCacheDetector(Set<String> cacheLocations) {
+    return CommandDetector.builder()
+      .with("rm")
+      .withOptionalRepeating(flagWithoutContainingNecessaryFlags)
+      .with(flagWithNecessaryRmOptions)
+      .withAnyFlag()
+      .with(cacheLocations)
+      .build();
+  }
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(RunInstruction.class, PackageInstallationCacheCheck::checkRunInstruction);
+  }
+
+  private static void checkRunInstruction(CheckContext ctx, RunInstruction runInstruction) {
+    List<ArgumentResolution> resolvedArgument = CheckUtils.resolveInstructionArguments(runInstruction);
+
+    SeparatedList<List<ArgumentResolution>, String> splitCommands = ArgumentResolutionSplitter.splitCommands(resolvedArgument);
+
+    List<CommandDetector.Command> sensitiveApkInstallCommands = new ArrayList<>();
+    List<CommandDetector.Command> sensitiveAptInstallCommands = new ArrayList<>();
```

Comment:
Are two different lists with two independent `forEach` required here or can all sensitive commands be collected into a single list? I don't think there is any difference in handling them after discovering.

Reply:
In `removeCacheCleanedInstallCommands` I remove all sensitive commands from these lists if there is a clean cache that is later. When we would collect all installs in a single list, we would lose the information if the install belongs to apk or apt. 
This would complicate some other parts of the code in such a way, that I don't feel it's worth it.